src/zkey/cca.c

```diff
@@ -587,3 +694,273 @@ int select_cca_adapter(struct cca_lib *cca, unsigned int card,
+			reason_code);
+
+	if (return_code != 0) {
+		print_CCA_error(return_code, reason_code);
+		return -EIO;
+	}
+
+	/* Generate secure key token */
+	memcpy(rule_array, "MASTER  ", CCA_KEYWORD_SIZE);
+	rule_array_count = 1;
+	cca->dll_CSNDPKG(&return_code, &reason_code, &exit_data_len, exit_data,
+			&rule_array_count, rule_array, &regen_data_len, NULL,
+			&skel_token_len, skel_token, trans_key, &ecc_token_len,
+			(unsigned char *)&ecc_token);
+
+	pr_verbose(verbose, "CSNDPKG (PKA Key Generate) "
+			"returned: return_code: %ld, reason_code: %ld", return_code,
+			reason_code);
+
+	if (return_code != 0) {
+		print_CCA_error(return_code, reason_code);
+		return -EIO;
+	}
+
+	memcpy(secure_key, ecc_token, ecc_token_len);
+	*seclen = ecc_token_len;
+
+	/* Extract public key from secure key token */
+	rc = ec_key_extract_public_cca(cca, ecc_token, ecc_token_len,
+				(unsigned char *)&buf, &buflen, verbose);
+	if (rc == 0 && buflen > 0) {
+		memcpy(public_key, buf, buflen);
+		*publen = buflen;
+	}
+
+	return 0;
+}
+
+/**
+ * Extract the public key out if the given secure key blob. A public key can
+ * only be obtained, if it is already contained in the token. The host lib
+ * does not re-calculate the public key from the given private key.
+ */
+int ec_key_extract_public_cca(struct cca_lib *cca, unsigned char *ecc_token,
+				unsigned int ecc_token_len, unsigned char *ecc_pub_token,
+				unsigned int *p_ecc_pub_token_len, bool verbose)
+{
+	unsigned char rule_array_buf[CCA_RULE_ARRAY_SIZE] = { 0 };
+	unsigned char buf[CCA_EC_KEY_VALUE_STRUCT_SIZE] = { 0 };
+	unsigned char exit_data[4];
+	long return_code = 0, reason_code = 0, exit_data_len = 0;
+	long rule_array_count = 0;
+	long buflen = sizeof(buf);
+	struct eccpubtoken *pubtok;
+
+	if (cca->dll_CSNDPKX == NULL) {
+		pr_verbose(verbose, "CCA host lib function CSNDPKX not "
+				"available but required for public key extract.");
+		return -ELIBACC;
+	}
+
+	/* Extract the public key from secure key token */
+	cca->dll_CSNDPKX(&return_code, &reason_code, &exit_data_len, exit_data,
+			&rule_array_count, rule_array_buf, (long *)&ecc_token_len, ecc_token,
+			&buflen, (unsigned char *)&buf);
+
+	pr_verbose(verbose, "CSNDPKX (PKA Public Key Extract) "
+			"returned: return_code: %ld, reason_code: %ld", return_code,
+			reason_code);
+
+	if (return_code != 0) {
+		print_CCA_error(return_code, reason_code);
+		return -EIO;
+	}
+
+	/* Copy result: public token comes after the CCA header */
+	pubtok = (struct eccpubtoken *)&buf[CCA_EC_HEADER_SIZE];
+	if (pubtok->curve_type == CCA_EC_CURVE_TYPE_PRIME) {
+		/* Remove indication for uncompressed key 0x04 */
+		memcpy(ecc_pub_token, pubtok->q + 1, pubtok->q_len - 1);
+		*p_ecc_pub_token_len = pubtok->q_len - 1;
+	} else {
+		memcpy(ecc_pub_token, pubtok->q, pubtok->q_len);
+		*p_ecc_pub_token_len = pubtok->q_len;
+	}
+
+	return 0;
+}
+
+int
+ec_key_clr2sec_cca(struct cca_lib *cca, int curve, unsigned int flags,
```

Comment:
int on same line ?

Reply:
Done. Yes, in the existing code the return type is often coded in a separate line, while I'm usually prefer to have it on the same line with the function name.